### PR TITLE
Dynamic programming strategy for determinant computations

### PR DESCRIPTION
**Pull request by haerski**
Implements the dynamic programming based algorithm by @xlni in C++. The algorithm is based on the cofactor method, with a few clever bookkeeping tricks to avoid unnecessary and repeat computations, especially for minors and exterior powers. See also the corresponding [Project wiki page](https://github.com/Macaulay2/M2/wiki/Project%3A-Faster-determinants-and-minors).

To use, pass the option `Strategy => Dynamic` to `determinant`, `minors`, or `exteriorPower`. Seems to work well with both dense and sparse matrices, although the caching introduces a bit of overhead compared to `Strategy => Cofactor` for very very sparse matrices.

```macaulay2
R = QQ[a_1..a_49]
M = genericMatrix(R,7,7)
elapsedTime exteriorPower(6,M, Strategy => Bareiss); -- slow
elapsedTime exteriorPower(6,M, Strategy => Cofactor); -- 1.91006 seconds elapsed
elapsedTime exteriorPower(6,M, Strategy => Dynamic); -- 0.659708 seconds elapsed

I = id_(R^500);
elapsedTime det(I, Strategy => Bareiss); -- slow
elapsedTime det(I, Strategy => Cofactor); -- 0.10984 seconds elapsed
elapsedTime det(I, Strategy => Dynamic); -- 0.225553 seconds elapsed

R = QQ[x,y,z]
setRandomSeed 42
N = random(R^{10:1}, R^{10:0}, Density => 0.2);
elapsedTime minors(9,N, Strategy => Bareiss); -- 8.91253 seconds elapsed
elapsedTime minors(9,N, Strategy => Cofactor); -- 6.53235 seconds elapsed
elapsedTime minors(9,N, Strategy => Dynamic); -- 1.7926 seconds elapsed
```

### CODE REVIEW

**Comment by mahrud** [comment]
Does this PR also resolve #487, by any chance?

**Comment by haerski** [comment]
Sorry for the delay, but I finally got around undrafting this PR. I added support for CTRL+C interruption, a bunch of additions to documentation, and ran it through clang-format.

**Comment by DanGrayson** [comment]
lgtm

@mikestillman -- can you take a look?

File: M2/Macaulay2/m2/multilin.m2 [hunk 1]
```diff
@@ -16,11 +16,13 @@ hasNoQuotients Ring := (R) -> true
 getMinorsStrategy := (R, opts) -> (
      bareiss := 0;  -- WARNING: these must match the engine!!
      cofactor := 1;
+     dynamic := 2;
      strat := if opts.?Strategy then opts.Strategy else null;
      if strat === global Bareiss then bareiss
      else if strat === global Cofactor then cofactor
+     else if strat === global Dynamic then dynamic
```
Comment (mikestillman) [inline M2/Macaulay2/m2/multilin.m2:23]: I suggest that "Dynamic" be the default.  Except I also suggest that if "Limit" or "First" is given, then use Cofactor.  When should Bareiss be the default?  I see very few cases where it is better...  Even when it is a bit better, it seems that Dynamic is only a bit worse.

File: M2/Macaulay2/e/det.cpp [hunk 2]
```diff
@@ -90,6 +91,102 @@ DetComputation::~DetComputation()
...
+                  flag = true;
+                  dynamic_cache[0][++nonzero] = {};
+                  row_lookup[i] = nonzero;
+                }
+              dynamic_cache[0][nonzero][{{i}, {j}}] = M->elem(i, j);
+            }
+        }
+    }
+  int n_nonzero_rows = dynamic_cache[0].size();
+  for (int minor_size = 1; minor_size < p; ++minor_size)
```
Comment (mikestillman) [inline M2/Macaulay2/e/det.cpp:116]: I'm a bit confused with the logic, and what is being automatically initialized.

**Review by mikestillman — APPROVED** [review]
I feel that this pull request should be approved.  In almost all cases of ```minors`` ` that I test, the new code is quite a bit faster than either Cofactor or Bareiss.  

I am suggesting some changes, and have some comments, for the next revision of this code.   But, as all the tests are passing, I say we pull it to get it into this release.

Thanks for the nice addition!

**Pull request merged**